# Matrix diagonalization in Python

In the Heisenberg formulation of quantum mechanics, an operator can represented as a **matrix** - specifically a hermitian matrix. Diagonalizing hermitian matrices therefore is crucial in condensed matter physics.

$$\mathbf{H}\Psi = \mathbf{E}\Psi$$
$$\Psi^T \mathbf{H} \Psi = \mathbf{E}$$

If there are $n$ basis vectors $\psi_i \forall i=1,...,N$, then $\mathbf{H}$ is an $n\times n$ hermitian matrix in this basis. The diagonal elements of the matrix $\mathbf{E}$ are the eigenvalues $\{E_i\}$.

In [1]:
import numpy as np
import matplotlib.pyplot as plt

__author__ = 'JL Euste'

In [2]:
import mercury as mr
app = mr.App(title='Matrix diagonalization',
             description='Introduction to matrix diagonalization in condensed matter physics',show_code=True)

mercury.App

Suppose we have the hamiltonian $H$ below in some arbitrary unit; this is actually the two-site Heisenberg hamiltonian in the spin basis $\{|\uparrow,\downarrow>,|\downarrow,\uparrow>, |\uparrow,\uparrow>,|\downarrow,\downarrow>\}$.

Our goal is to obtain the full spectrum. We can calculate the eigenstates and eigenvalues (i.e. energy values) by diagonalizing H.

In [3]:
H = np.array([[-1/4,1/2,0,0],
              [1/2,-1/4,0,0], 
              [0,0,1/4,0], 
              [0,0,0,1/4]])
H

array([[-0.25,  0.5 ,  0.  ,  0.  ],
       [ 0.5 , -0.25,  0.  ,  0.  ],
       [ 0.  ,  0.  ,  0.25,  0.  ],
       [ 0.  ,  0.  ,  0.  ,  0.25]])

We diagonalize $H$ using `eigh` from the `linalg` library in `scipy`. This gives the eigenvalues in *ascending* order and the *normalized* eigenvectors. Check the documentation for more details.

In [4]:
from scipy.linalg import eigh
eigvals, eigvecs = eigh(H)

In [5]:
eigvals

array([-0.75,  0.25,  0.25,  0.25])

The `eigvals` contains the eigenvalues in ascending order. We see that ground state (GS) energy is -3/4 while the excited state is triply degenerate with energy 1/4. But which state correponds to which energy level? We now look at the eigenvectors `eigvecs`.

In [6]:
eigvecs

array([[ 0.70710678,  0.70710678,  0.        ,  0.        ],
       [-0.70710678,  0.70710678,  0.        ,  0.        ],
       [ 0.        ,  0.        ,  1.        ,  0.        ],
       [ 0.        ,  0.        ,  0.        ,  1.        ]])

Each column is an eigenvector corresponding to the eigenvalue of the same index so the first column gives the coefficients of the basis states in the GS while the last three columns in the excited state. We write a short function to neatly display the eigenvectors and corresponding eigenvalues $E$.

In [7]:
def display_eigen(eigen):
  '''Display each eigenvector with its corresponding eigenvalue.'''
  for i in range(len(eigen[0])):
    print(f'eigenvector {eigen[1][:,i]}  with E = {eigen[0][i]:.3f}')

In [8]:
display_eigen([eigvals,eigvecs])

eigenvector [ 0.70710678 -0.70710678  0.          0.        ]  with E = -0.750
eigenvector [0.70710678 0.70710678 0.         0.        ]  with E = 0.250
eigenvector [0. 0. 1. 0.]  with E = 0.250
eigenvector [0. 0. 0. 1.]  with E = 0.250


The eigenvectors are eigenstate which we can write in terms of the basis vectors. Note: $0.70710678=\frac{1}{\sqrt{2}}$

For example, the GS is
$|GS> = \frac{1}{\sqrt{2}}|\uparrow,\downarrow> - \frac{1}{\sqrt{2}}|\downarrow,\uparrow> + 0|\uparrow,\uparrow> + 0|\downarrow,\downarrow> = \frac{1}{\sqrt{2}}(|\uparrow,\downarrow> - |\downarrow,\uparrow>) $

Therefore, the ground state is the singlet while the excited states are the triplet states:

$|\psi> = \frac{1}{\sqrt{2}}( |\downarrow,\uparrow> + |\uparrow,\downarrow>) $;   
$|\psi> = |\uparrow,\uparrow>$;   
$|\psi> = |\downarrow,\downarrow>$.


In [9]:
eigvecs.T @ H @ eigvecs

array([[-7.50000000e-01,  1.11022302e-16,  0.00000000e+00,
         0.00000000e+00],
       [ 1.38777878e-16,  2.50000000e-01,  0.00000000e+00,
         0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  2.50000000e-01,
         0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         2.50000000e-01]])

***

ADDENDUM: We use scipy's `eigh` function instead of that from numpy because the former seems more reliable (for reasons I'm not sure). Below we compare the diagonalization of the `sample` matrix using numpy and scipy.

In [10]:
sample = np.array([[ 0. , -0.5,  0. ,  0. ,  0. , -0.5,  0. ,  0. ],
       [-0.5,  0. , -0.5,  0. ,  0.5,  0. , -0.5,  0. ],
       [ 0. , -0.5,  0. , -0.5,  0. ,  0.5,  0. , -0.5],
       [ 0. ,  0. , -0.5,  0. ,  0. ,  0. ,  0.5,  0. ],
       [ 0. ,  0.5,  0. ,  0. ,  0. ,  0.5,  0. ,  0. ],
       [-0.5,  0. ,  0.5,  0. ,  0.5,  0. ,  0.5,  0. ],
       [ 0. , -0.5,  0. ,  0.5,  0. ,  0.5,  0. ,  0.5],
       [ 0. ,  0. , -0.5,  0. ,  0. ,  0. ,  0.5,  0. ]])
np.set_printoptions(suppress=True,precision=2) #for neat formatting of array elements

*Numpy*

In [11]:
np_eigvals,np_eigvecs = np.linalg.eigh(sample)
display_eigen([np_eigvals,np_eigvecs])

eigenvector [ 0.19 -0.21 -0.19  0.22 -0.19  0.6  -0.63  0.22]  with E = -1.000
eigenvector [ 0.46  0.63  0.08 -0.09 -0.46  0.29  0.26 -0.09]  with E = -1.000
eigenvector [-0.    0.23  0.68  0.44  0.   -0.24 -0.2   0.44]  with E = -1.000
eigenvector [ 0.71 -0.    0.   -0.    0.71 -0.    0.   -0.  ]  with E = -0.000
eigenvector [-0.    0.   -0.   -0.71  0.    0.    0.    0.71]  with E = -0.000
eigenvector [ 0.    0.1  -0.59  0.49  0.   -0.1   0.39  0.49]  with E = 1.000
eigenvector [-0.5  0.5  0.   0.   0.5  0.5 -0.   0. ]  with E = 1.000
eigenvector [-0.   -0.49  0.39  0.1  -0.    0.49  0.59  0.1 ]  with E = 1.000


*Scipy*

In [12]:
sp_eigvals,sp_eigvecs = eigh(sample)
display_eigen([sp_eigvals,sp_eigvecs])

eigenvector [ 0.   0.  -0.5 -0.5  0.   0.   0.5 -0.5]  with E = -1.000
eigenvector [ 0.  -0.5 -0.5  0.   0.   0.5 -0.5  0. ]  with E = -1.000
eigenvector [-0.5 -0.5  0.   0.   0.5 -0.5  0.   0. ]  with E = -1.000
eigenvector [-0.71  0.   -0.    0.   -0.71  0.   -0.    0.  ]  with E = 0.000
eigenvector [ 0.    0.    0.   -0.71 -0.    0.   -0.    0.71]  with E = 0.000
eigenvector [ 0.   0.   0.5 -0.5 -0.   0.  -0.5 -0.5]  with E = 1.000
eigenvector [ 0.  -0.5  0.5  0.  -0.   0.5  0.5  0. ]  with E = 1.000
eigenvector [ 0.5 -0.5 -0.   0.  -0.5 -0.5 -0.   0. ]  with E = 1.000


Notes on comparing scipy and numpy eigendecompositions:
- Eigenvalues are the same and order in ascending order by default.
- Eigenvectors are both normalized and correct.
- Eigenvectors generated from scipy, however, are "simpler". For instance, the first eigenvector from numpy is $[ 0.19,-0.21,-0.19,0.22,-0.19,0.6,-0.63,0.22]$ while the one from scipy can be written as $[0,0,-\frac{1}{2},-\frac{1}{2},0,0,\frac{1}{2},-\frac{1}{2}]$ 

***